### DISCUSSION TITLE
[RFC] What is the concern of creating modules as bundles?

### DISCUSSION THREAD

**Discussion by fico7489** [symfony/symfony#53997] [Q&A]
### Description

Symfony official documentation says:

    "Don't Create any Bundle to Organize your Application Logic"

https://symfony.com/doc/current/best_practices.html

And:

    In Symfony versions prior to 4.0, it was recommended to organize your own application code using bundles. This is no longer recommended and bundles should only be used to share code and features between multiple applications.

https://symfony.com/doc/current/bundles.html

Lets assume that I have 10 modules (which are also Bundles) in my app: "Company Module", "User Module", "Accounting Module", etc.

Bundles(Modules) are not separate Composer packages; instead, they are committed as part of the app code, and each Bundle(Module) is included in config/bundles.php.

Each module is created as a bundle with its own configuration, and the average size of each configuration is around 100 lines.

For example each of this file have in average 100 lines of code:

`Module\[MODULE_NAME]\Resource\config\services.yaml`

Also some of the modules(bundles) have routes config:

`Module\[MODULE_NAME]\Resource\config\routes.yaml`

Some of them workflow config:

`Module\[MODULE_NAME]\Resource\config\workflow.yaml`

etc.

And that is perfectly fine by me. I want to put module configuration inside each module folder, not in config folder in the root (config/services.yaml).

What is alternative? Put everything from each module configuration.yaml to config/services.yaml and have messy file with 1000 lines of configuration?

Or put configuration inside each module (which is not bundle) and then import it with inside config/services.yaml with:

```
imports:
    - { resource: Module\[MODULE_NAME]\Resource\config\services.yaml }
    - { resource: Module\[MODULE_NAME2]\Resource\config\services.yaml }
    - { resource: Module\[MODULE_NAME3]\Resource\config\services.yaml }
    ...

```
That is also a little bit messy for me.

When I need to use templates within a module that isn't a bundle, I have to place the templates inside a 'templates' folder at the root.
```

templates/MODULE_NAME
templates/MODULE_NAME2
templates/MODULE_NAME3
...
```

This is already the third instance where I repeatedly find myself creating ten folders for each module. There are additional situations, such as translations, where this approach should also be applied, potentially leading to more than 10 instances where this task needs to be performed in the same manner.

Isn't it easier to create a module as a bundle and keep all its components inside that bundle, rather than creating ten folders for each bundle in multiple locations?

I am perfectly happy when "**UserModule**" code is only in folder for "**UserModule**" not everywhere around.

What is a general concern of creating a modules as bundles (Organize your Application Logic in bundles)?

**Comment by asispts** [comment]
IMHO, you're over-engineering your project.

**Comment by yceruto** [comment]
I still see value in organizing large projects using independent and encapsulated modules, and that's not a problem. The thing is that bundles have an opinionated structure, which is fine for the purpose it was designed for, but it might not fit your project requirements.

For instance, large projects using a layered architecture usually organize the code differently, and it's not possible to adapt using the bundle approach.

On the other hand, bundles are meant to be isolated packages, so your dependencies come from vendor/external libraries only. In your project, you'll need inter-module dependencies, e.g., a direct relationship between UserModule and AccountingModule. Coupling your bundle with other bundles in your own project defeats the bundle's purpose, or it overcomplicates the project development if you want to make them completely isolated from each other.

Those, I think, are some reasons why Symfony doesn't recommend organizing your project using bundles.

**Comment by yceruto** [comment]
However, you could use the [DI extension](https://github.com/symfony/symfony/blob/db21ee4abecfe3bd403c7e655753e2e087b8582a/src/Symfony/Component/DependencyInjection/Extension/AbstractExtension.php) approach if you want to achieve more/less the same but with the freedom of creating your own module conventions to load config, services, etc.

**Comment by fico7489** [comment]
@asispts, I did not design the structure of the project; I started working on it a few years after the project began. Modules are currently created as bundles, and everything is going well with the project development. I am now confused; if I were to convert modules from bundles to non-bundles, I believe everything would become messy, as I explained above in the example. That is the reason why I am asking this question.

**Comment by mbabker** [comment]
Everything you say about not liking building an app without bundles (using paths like `templates/<MODULE>`, importing `@Module/Resources/config/services.yaml`, routing configs split across modules, translations split across modules, etc.) already exists in the bundle structure.  The difference is that the perceived messy filesystem structure has some resemblance of a normalized structure because of the standardized paths that bundles ask for, and you don't have a "main" resource (like the `config/services.yaml` in a new app) responsible for importing all of those module resources.
